### ISSUE TITLE
how to use Controls? persistent search

### ISSUE THREAD

**Issue by martinrm77** [cannatag/ldap3#194]
First, ldap3 is awesome.

Im trying to figure out how to use a persistent search control by creating a control object with
persistent_search_control = ldap3.protocol.controls.build_control('2.16.840.1.113730.3.4.3', True, 'some value I dont know how to supply')
How do you supply the controlValue ? its not a string.
The RFC (https://tools.ietf.org/html/draft-ietf-ldapext-psearch-00) says the control needs a sequence like this:
PersistentSearch ::= SEQUENCE {
                   changeTypes INTEGER,
                   changesOnly BOOLEAN,
                   returnECs BOOLEAN
           }

How do you supply this value in something that it can BER encode?

And I was thinking I can supply it to the search function and fire away.

**Comment by martinrm77** [comment]
Ok, I think I got the jist of it, but it just wont work - search wont take my built controls sequence.
ldap3.core.exceptions.LDAPControlsError: controls must be a sequence

and my controls is 
Control(componentType=NamedTypes(NamedType('controlType', LDAPOID()), DefaultedNamedType('criticality', Criticality('False')), OptionalNamedType('controlValue', ControlValue()))).setComponents(LDAPOID(b'2.16.840.1.113730.3.4.3'), Criticality('True'), ControlValue(hexValue='3009020107010100010100'))'

which is a Sequence! even python 3.5 says so.
>>> isinstance(persistent_search_control, Sequence)
True
>>> persistent_search_control
Control(componentType=NamedTypes(NamedType('controlType', LDAPOID()), DefaultedNamedType('criticality', Criticality('False')), OptionalNamedType('controlValue', ControlValue()))).setComponents(LDAPOID(b'2.16.840.1.113730.3.4.3'), Criticality('True'), ControlValue(hexValue='3009020107010100010100'))
>>> type(persistent_search_control)
<class 'ldap3.protocol.rfc4511.Control'>

Any help you can give me?

**Comment by martinrm77** [comment]
Well, persistent search needs a complex value - It needs the sequence in the RFC.

I think I got farther simply by putting the object in a tuple... (enclosing with (,) ) it does say options, not option, so I guess its semantically correct.

I guess I need to make an ASYNC connection and check the response object.

When I get it working I would love to do a pull request for a nice persistent search control object, to easy put in the correct options and a documented example. I think its a fantastic thing to do persistent searches, its nice and realtime.. if it works.

**Comment by cannatag** [comment]
I would be happy to help. Which kind of ldap server are you trying to connect? I'll try to replicate in my lab.

Giovanni

**Comment by martinrm77** [comment]
I'm connecting to a ds389 - an updated RHEL IPA installation.
Well, this is kinda what I got until now, doesnt give any output thou :

```
class persistent_search_control_value(Sequence):
        componentType = NamedTypes(NamedType('changeTypes', Integer()),
                                   NamedType('changesOnly', Boolean()),
                                   NamedType('returnECs', Boolean()),
                                   )

pvalue = persistent_search_control_value()
pvalue['changeTypes'] = 15
pvalue['changesOnly'] = False
pvalue['returnECs'] = False
persistent_search_control = ldap3.protocol.controls.build_control('2.16.840.1.113730.3.4.3', True, pvalue)

ldapconn = ldap3.Connection('ipaserver01.local', client_strategy='ASYNC', user='uid=user,cn=users,cn=accounts,dc=ipa,dc=ku,dc=dk', password='secretpass', auto_bind=True)
msg = ldapconn.search('cn=dns,dc=ipa,dc=ku,dc=dk','(objectclass=idnszone)', controls=(persistent_search_control,))
resp=ldapconn.get_response(msg, timeout=900)
```
I think it looks fine, but it should be watching for all changes to new dns zones. Nothing yet..

**Comment by cannatag** [comment]
Please have a look at https://github.com/cannatag/ldap3/blob/master/ldap3/extend/microsoft/dirSync.py

It performs an Active Directory Sync, and keeps receiving all changes in the Active Directory Tree. I think you could use the same structure to achieve the persistent search result.

Anyway I could work on this in the weekend.

**Comment by martinrm77** [comment]
hmm.. I can see its getting all the right data through the network, but get_response times out. How does there data come back?

Excerpt from debug log:
```
<<LDAPMessage:
<< messageID=2
<< protocolOp=ProtocolOp:
<<  searchResEntry=SearchResultEntry:
<<   object=b'idnsname=125.38.192.in-addr.arpa.,cn=dns,dc=ipa,dc=ku,dc=dk'
<<   attributes=PartialAttributeList:
DEBUG:ldap3:NETWORK:received 1 ldap message via <ldap://ipa01pl.ipa.ku.dk:389 - cleartext - user: uid=ns_user,cn=users,cn=accounts,dc=ipa,dc=ku,dc=dk - not lazy - bound - open - <local: 192.38.110.174:39320 - remote: 10.77.40.200:389> - tls not started - listening - AsyncStrategy - pyasn1 decoder>
DEBUG:ldap3:EXTENDED:ldap message received via <ldap://ipaserver01:389 - cleartext - user: uid=user,cn=users,cn=accounts,dc=ipa,dc=ku,dc=dk - not lazy - bound - open - <local: 192.38.110.174:39320 - remote: 10.77.40.200:389> - tls not started - listening - AsyncStrategy - pyasn1 decoder>:
<<LDAPMessage:
<< messageID=2
<< protocolOp=ProtocolOp:
<<  searchResEntry=SearchResultEntry:
<<   object=b'idnsname=wild.local.,cn=dns,dc=ipa,dc=ku,dc=dk'
<<   attributes=PartialAttributeList:

```

**Comment by martinrm77** [comment]
Ok, this is not that easy.

It looks like Im getting lots of responses, and they are probably also saved in the receivers response buffer, but never returned to me, because there is never a searchResDone=SearchResultDone result.

This is what I gather from reading the client strategies, I tried SYNC, ASYNC, REUSEABLE (which looks quite fantastic to me), but none return anything before the result is done from the server.

Can you verify?

for example in strategy/async.py line 204:
`responses = self._responses.pop(message_id) if message_id in self._responses and self._responses[message_id][-1] == RESPONSE_COMPLETE else None`

Im not well versed in this low level programming, so I dont know how to fix it, but there will be no persistent search before this is addressed.

Best regards,
Martin

**Comment by cannatag** [comment]
Sure. I'll work on this in the weekend. 

Bye,
Giovanni

**Comment by cannatag** [comment]
Hi Martin,
please test the following code with the latest ldap3 in dev:

```
from ldap3 import Server, Connection, ASYNC_STREAM
s = Server('myserver')
c = Connection(s, 'cn=admin,o=resources', 'password', client_strategy=ASYNC_STREAM)
c.stream = open('myfile.log', 'w+)
p = c.extend.standard.persistent_search()
```

now the persistent search is running. Each modification is recorded in the log in LDIF-CHANGE format, with the event type, event time and the modified dn (if available) as comments.

In the persistent_search() method you can use the same parameter of a standard search. It also accepts some additional parameters specific of the persistent search:

```
    def persistent_search(self,
                          search_base='',
                          search_filter='(objectclass=*)',
                          search_scope=SUBTREE,
                          dereference_aliases=DEREF_NEVER,
                          attributes=ALL_ATTRIBUTES,
                          size_limit=0,
                          time_limit=0,
                          changes_only=True,
                          show_additions=True,
                          show_deletions=True,
                          show_modifications=True,
                          show_dn_modifications=True,
                          notifications=True,
                          controls=None
                          ):
```
If you don't pass any parameters the search should be globally applied in your LDAP server.

If you want to stop the persistent search you can use ``p.stop()``. Use ``p.start()`` to start it again.

If you don't provide a stream (a file to write to), a StringIO object is used. You can use it as a standard file or get the value of the StringIO object with ``c.stream.getvalue()``.

For example an output from my test suite is the following:
```
# 2016-07-10T23:34:41.616615
# add
dn: cn=[71973491]modify-dn-1,o=test
objectClass: inetOrgPerson
objectClass: organizationalPerson
objectClass: Person
objectClass: ndsLoginProperties
objectClass: Top
sn: modify-dn-1
cn: [71973491]modify-dn-1
ACL: 2#subtree#cn=[71973491]modify-dn-1,o=test#[All Attributes Rights]
ACL: 6#entry#cn=[71973491]modify-dn-1,o=test#loginScript
ACL: 2#entry#[Public]#messageServer
ACL: 2#entry#[Root]#groupMembership
ACL: 6#entry#cn=[71973491]modify-dn-1,o=test#printJobConfiguration
ACL: 2#entry#[Root]#networkAddress

# 2016-07-10T23:34:41.888506
# modify dn
# previous dn: cn=[71973491]modify-dn-1,o=test
dn: cn=[71973491]modified-dn-1,o=test
objectClass: inetOrgPerson
objectClass: organizationalPerson
objectClass: Person
objectClass: ndsLoginProperties
objectClass: Top
sn: modify-dn-1
cn: [71973491]modified-dn-1
ACL: 2#subtree#cn=[71973491]modified-dn-1,o=test#[All Attributes Rights]
ACL: 6#entry#cn=[71973491]modified-dn-1,o=test#loginScript
ACL: 2#entry#[Public]#messageServer
ACL: 2#entry#[Root]#groupMembership
ACL: 6#entry#cn=[71973491]modified-dn-1,o=test#printJobConfiguration
ACL: 2#entry#[Root]#networkAddress

# 2016-07-10T23:34:41.929022
# delete
dn: cn=[71973491]modified-dn-1,o=test
objectClass: inetOrgPerson
objectClass: organizationalPerson
objectClass: Person
objectClass: ndsLoginProperties
objectClass: Top
sn: modify-dn-1
cn: [71973491]modified-dn-1
ACL: 2#subtree#cn=[71973491]modified-dn-1,o=test#[All Attributes Rights]
ACL: 6#entry#cn=[71973491]modified-dn-1,o=test#loginScript
ACL: 2#entry#[Public]#messageServer
ACL: 2#entry#[Root]#groupMembership
ACL: 6#entry#cn=[71973491]modified-dn-1,o=test#printJobConfiguration
ACL: 2#entry#[Root]#networkAddress
```

I've tested it against a NetIQ eDirectory. Not sure if it works on different servers.

Is this helpful for your intended use? An alternative should be to store the events in a queue and let your code manage it. 

Let me know what you think.

Bye,
Giovanni

**Comment by martinrm77** [comment]
This is all very nice, I tested it against our RHEL IPA installation on ds389, works like a charm.

But, I would much rather get the responses with get_response() and the response_dict - I dont want to parse the LDIF again. Is it possible to get the response dictionaries like with ASYNC strategy?

I simply want to process each record as they appear, so maybe passing a function that gets each response?

/Martin

**Comment by cannatag** [comment]
Hi, the problem with get_response() is that the response never ends! I'll try to hack the method to return what is in the response queue. But then is up to your code to check it, probably in an asynchronous task, requesting for the same message id repeatedly. A different way to accomplish this could be passing a callback that calls a function in your code every time an event is received, but this increase the code complexity.

What do you think?

**Comment by martinrm77** [comment]
Preferably both.

Instead of you getting dicts from _responses and formatting them in LDIF and putting them on .stream - what about a .stream_responses where you put in the response dicts?

Passing a callback function that would be executed with each response as they come in would be clean and efficient - the fastest way to get to the responses. I think this solution is best for users of the module, otherwise I have to poll either .stream or .stream_responses - which would also work off course.

Thanks for the support.

/Martin

**Comment by cannatag** [comment]
Hi Martin, please find in dev an updated version with returned entries and callback.

If you call the persistent_search() method with ``straming=False`` you can get the modified entries with the ``p.next()`` method. Each call to p.next() will return one event, with the extended control already decoded (as dict values) if available. I had to create a new method because get_response() never sees the end of the search() operation so it would never return the entries. 

If you call the persistent_search() method with ``callback=myfynction`` (where myfunction is a callable, including lambda accepting, a dict as parameter)  your function will be called for each event received in the persistent serach. The function will be called in the same thread of the persistent search, so it should not block.

I'm not really satisfied with both of this solution. WIth the former you must keep calling the p.next() method every n seconds to be sure to not accumulate too much events in the queue. With the latter this problem is solved, but a bad behaviour of myfunction could stop the receiving of persistent search.

Please experiment and tell me what do you think about them.

Bye,
Giovanni

**Comment by cannatag** [comment]
released in 1.4.0
